### PR TITLE
Fix redefined & overloaded methods on fast_path

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This unblocks #5808.

### Commit summary

- **pre-work: Add new UniqueNameKind::MangleRenameOverload** (9f808f349)

  The way that overloads works fundamentally requires mangle renaming the
  original method, so that there is still one method in the symbol table
  with all the ArgInfo's, that we can selectively copy onto each (derived)
  overload method.

  That means that it definitely needs the logic that mangleRenameSymbol
  does to rename things and erase things from the `members` vector.

  But I want to do it in a way that's more predictable, so that it's
  easier to map a name that was mangled in this way back to the original
  method, as justified in the next commit.

- **Add mangleRenameForOverload method** (95a89ec69)

  Now, instead of calling mangleRenameSymbol in resolver, we call
  mangleRenameForOverload. This lets us map back and forth between a
  method that was mangled only for the purpose of overloading (ignoring
  methods that have been mangle renamed in an attempt to recover from
  method redefinitions).

  This involves a slight trick in `lookupMethodSymbolWithHash`, because in
  addition to walking up the chain of MangleRename unique names, it must
  also look for a `MangleRenameOverload` at that step. In a picture:

  Before:

  ```
  <U example>
      │
      ▼
  <M <U example> $1>
      │
      ▼
  <M <U example> $2>
      │
      ▼
      ⋮
  ```

  Now:

  ```
  <U example>
      │
      ▼
  <M <U example> $1>  ─▶  <V <M <U example> $1> $1>
      ┌────────────────────────────┘
      ▼
  <M <U example> $2>  ─▶  <V <M <U example> $2> $1>
      ┌────────────────────────────┘
      ▼
      ⋮
  ```

  With this in place, we can relax/rework some assumptions in
  enterNewMethodOverload, namely: the method overload might already exist
  (because we're on the fast path). The new `ENFORCE`'s all combine to
  basically say that if there's already an method overload with the
  current name, that it must match exactly the arity we're expecting.

- **Two other edge cases** (5ccd7fa7a)

- **Fix a bunch of tests** (3c76998be)

### Test plan

Existing tests become good, and added a new test

### CODE REVIEW

File: namer/namer.cc [hunk 1]
```diff
@@ -853,6 +853,17 @@ class SymbolDefiner {
...
     core::MethodRef insertMethod(core::MutableContext ctx, const core::FoundMethod &method) {
         auto symbol = defineMethod(ctx, method);
+        auto name = symbol.data(ctx)->name;
+        if (name.kind() == core::NameKind::UNIQUE &&
+            name.dataUnique(ctx)->uniqueNameKind == core::UniqueNameKind::MangleRenameOverload) {
+            // These name kinds are only created in resolver, which means that we must be running on
+            // the fast path with an existing GlobalState.
+            // When modifyMethod is called later, it won't be able to find the correct method entry.
+            // Let's leave the method visibility what it was.
+            // TODO(jez) After #5808 lands, can we delete this check?
```
Comment (jez) [inline namer/namer.cc:863]: yes, we can

https://github.com/sorbet/sorbet/pull/5808/commits/d8ba7db6799bd29cd339a9df6ab21fe3ef6be569

File: test/testdata/lsp/fast_path/overloads_test.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# typed: true
+
+class A
```
[the remark this answers is not in the document]
Comment (jez) [reply test/testdata/lsp/fast_path/overloads_test.rb:3]: yeah this test started as a test that i wrote on #5808 and backported to this PR. i discovered when I was presented with merge conflicts rebasing #5808 to be on top of this change that somehow in this branch i removed the fast/slow path assertion. i didn't feel like retroactively fixing this PR, rebasing, and fixing conflicts again—instead i opted to just put the correct assertion in #5808.

File: core/GlobalState.cc [hunk 3]
```diff
@@ -1627,9 +1653,14 @@ void GlobalState::mangleRenameSymbol(SymbolRef what, NameRef origName) {
...
-        name = freshNameUnique(UniqueNameKind::MangleRename, origName, collisionCount++);
-    } while (ownerData->findMember(*this, name).exists());
+    if (kind == UniqueNameKind::MangleRename) {
+        do {
+            name = freshNameUnique(UniqueNameKind::MangleRename, origName, collisionCount++);
+        } while (ownerData->findMember(*this, name).exists());
+    } else {
+        ENFORCE(kind == UniqueNameKind::MangleRenameOverload);
+        name = freshNameUnique(UniqueNameKind::MangleRenameOverload, origName, 1);
+    }
```
[the remark this answers is not in the document]
Comment (jez) [reply core/GlobalState.cc:1666]: ah i see, i can change that

[the remark this answers is not in the document]
Comment (jez) [reply core/GlobalState.cc]: @froydnj lmk if you like the comment i added. happy to improve if it's not what you were looking for

**Review by froydnj — APPROVED** [review]
It still makes sense to me that we ought to be able to `ENFORCE` the non-existence of the newly-mangled name (I see no testsuite failures with such an `ENFORCE`), but meh.

**Comment by jez** [comment]
From discussion:

>ok interesting thanks for pushing on this
>
> so the code in namer uses the arityHash to lookup the matching method regardless of name, which means that on the fast path namer will resolve the method definition to a Symbol that has the MangleRenameOverload name kind
>
> i was thinking (mistakenly) that resolver would then call mangleRenameForOverload on an already overloaded name and return the existing one, but that's not the case—resolver only calls mangleRenameForOverload if it would be needed, which is why the ENFORCE passes
>
> So it is actually an invariant that regardless of which branch we take (either the MangleRename case or the MangleRenameOverload case), once the if finishes, the ENFORCE you've written is true. Either it's true because it was the post condition of the loop, or it's true because the resolveSigs call site upholds it.
